### PR TITLE
Implemented face-based Dirichlet boundary condition in compositional solver

### PR DESCRIPTION
**Pull request by francoishamon**
This PR implements the isothermal and thermal Dirichlet face boundary conditions in the compositional solver. This is needed to extend [this wellbore example](https://geosx-geosx.readthedocs-hosted.com/en/latest/docs/sphinx/advancedExamples/validationStudies/verticalPoroElastoPlasticWellbore/Example.html) to CO2 injection. The PR also add the missing mesh body loops (`forMeshTargets`) in the functions validating/applying the boundary conditions.

In the second commit, the PR also replaces the fields `dPhaseVolFraction_dPres`, `dPhaseVolFraction_dTemp`, `dPhaseVolFraction_dComp` with the aggregate `dPhaseVolFraction`. The same is done for the phase mobility. This is needed to compile the face boundary condition kernel (otherwise, `formal parameter space overflow`). I am sorry there are so many modified files as a result.

Rebaseline needed in `deadoil_3ph_staircase_3d` and `co2_thermal_2d` that now have face boundary conditions. Rebaseline also needed for all the compositional integrated tests because of the new fields `dPhaseVolFrac` and `dPhaseMob`.

Rebaseline done in https://github.com/GEOSX/integratedTests/pull/240

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp [hunk 1]
```diff
@@ -628,6 +625,315 @@ void CompositionalMultiphaseFVM::updatePhaseMobility( ObjectManagerBase & dataGr
...
+  return bcConsistent;
+}
+
+namespace internal
+{
+string const faceBcLogMessage = string( "CompositionalMultiphaseFVM {}: at time {}s, " )
+                                + string( "the <{}> boundary condition '{}' is applied to the face set '{}' in '{}'. " )
+                                + string( "\nThe total number of target faces (including ghost faces) is {}. " )
+                                + string( "\nNote that if this number is equal to zero, the boundary condition will not be applied on this face set." );
+}
```
Comment (klevzoff) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:771]: I'd generally suggest the following (message text unchanged, only code formatting). I hope uncrustify doesn't mind. In future versions `{fmt}` will require the format string to be `constexpr`, so the `string` one might not work. I also removed the namespace name to avoid external linking for the symbol.
```suggestion
namespace
{
char const faceBcLogMessage[] =
  "CompositionalMultiphaseFVM {}: at time {}s, "
  "the <{}> boundary condition '{}' is applied to the face set '{}' in '{}'.\n" 
  "The total number of target faces (including ghost faces) is {}.\n"
  "Note that if this number is equal to zero, the boundary condition will not be applied on this face set.";
}
```

**Review by klevzoff — APPROVED** [review]
Nice! Makes me think this is how derivatives should've been stored from the beginning. We live, we learn :)

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp [hunk 2]
```diff
@@ -1396,157 +1391,156 @@ void CompositionalMultiphaseBase::applySourceFluxBC( real64 const time,
...
+                                               MeshLevel & mesh,
+                                               arrayView1d< string const > const & )
   {
-    // Check whether pressure has already been applied to this set
-    string const & subRegionName = subRegion.getName();
-    string const & regionName = subRegion.getParent().getParent().getName();
+    // maps to check consistent application of BC
+    map< string, map< string, map< string, ComponentMask< MAX_NC > > > > bcPresCompStatusMap; // check that pressure/comp are
+                                                                                              // present/consistent
+    map< string, map< string, set< string > > > bcTempStatusMap; // check that temperature is present/consistent
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp]: Done, I added some comments to document the mappings in the two validation functions.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp (hunk 1 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: Done, I put all the log messages that were in this format (namespace + string) in the flow solvers to the new format.

File: inputFiles/thermalMultiphaseFlow/co2_thermal_2d.xml [hunk 3]
```diff
@@ -187,27 +192,28 @@
 
     <FieldSpecification
       name="sourcePressure"
-      setNames="{ source }"     
-      objectPath="ElementRegions/region/cb"
+      logLevel="1"
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply inputFiles/thermalMultiphaseFlow/co2_thermal_2d.xml:195]: Yes, I prefer to keep the log levels to 1 in some integrated tests, because newcomers may modify these existing tests to construct their own xml files, and I think it is good to start with the log levels to 1 to avoid mistakes.

File: src/coreComponents/constitutive/solid/PorousSolid.hpp [hunk 4]
```diff
@@ -209,13 +208,13 @@ class PorousSolidUpdates : public CoupledSolidUpdates< SOLID_TYPE, BiotPorosity,
     {
       // Compute mixture density
       real64 fluidTotalMassDensity = fluidPhaseSaturation( 0 ) * fluidPhaseMassDensity( 0 );
-      real64 dFluidTotalMassDensity_dPressure = dFluidPhaseSaturation_dPressure( 0 ) * fluidPhaseMassDensity( 0 )
+      real64 dFluidTotalMassDensity_dPressure = dFluidPhaseSaturation( 0, Deriv::dP ) * fluidPhaseMassDensity( 0 )
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/constitutive/solid/PorousSolid.hpp:211]: Yes, maybe now is a good time to try this approach in a dedicated PR to see if we encounter some roadblocks along the way.

For your first question, I think this is consistent with what we do elsewhere. For a derivative of say, `phaseVolFraction` in a subregion, we do:
```dPhaseVolFrac[elementIndex][phaseIndex][derivativeIndex]```
so it is the same here, except that we consider a slice (the element index does not appear), and the first phase (index 0)

**Comment by francoishamon** [comment]
This PR requires a rebaseline, I am going to do it in a few hours
